### PR TITLE
Fix overflow in conversion from uvalue to svalue

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Resolves: #732
Resolves: #766
Resolves: #767 
Resolves: #768

Invoke the overflow_* APIs to handle cases where the value would have overflowed.

## Summary by CodeRabbit

- **New Features**
	- Enhanced handling of arithmetic operations, including left and right shifts.
	- Introduced a method for sign extension of integers.
	- Improved error handling for arithmetic operations.

- **Tests**
	- Added multiple test cases to validate that both signed and unsigned right shifts, as well as left shifts, by zero are idempotent, ensuring the original value remains unchanged across various scenarios.

### CODE REVIEW

File: src/crab/ebpf_domain.cpp [hunk 1]
```diff
@@ -2374,7 +2374,8 @@ void ebpf_domain_t::lshr(const Reg& dst_reg, int imm, int finite_width) {
             }
         }
         m_inv.set(dst.uvalue, interval_t{lb_n, ub_n});
-        if (ub_n.narrow<int64_t>() >= lb_n.narrow<int64_t>()) {
+        if ((ub_n.fits<int64_t>() && lb_n.fits<int64_t>()) &&
```
Comment (dthaler) [inline src/crab/ebpf_domain.cpp:2377]: If finite_width is 64 and imm is 0, then [lb,ub] can be something that would be (say) [-2,-2] as a signed number.
That is legal and should not be havoced.  So I think this should use `fits_cast_to` not `fits`.

File: test-data/shift.yaml [hunk 2]
```diff
@@ -1104,3 +1104,167 @@ post:
   - r2.type=number
   - r2.svalue=32
   - r2.uvalue=32
+
+---
+test-case: check that unsigned right shift by 0 is idempotent - svalue out of range
+
+pre:
+  - r0.type=number
+  #  This is 0x8000000000000000, which is out of range for a signed 64-bit number.
```
Comment (dthaler) [inline test-data/shift.yaml:1113]: It's not out of range, it's -9223372036854775808 as shown in the next line.

File: test-data/shift.yaml [hunk 3]
```diff
@@ -1104,3 +1104,167 @@ post:
...
+
+messages:
+  - "1:2: Code is unreachable after 1:2"
+
+---
+test-case: check that unsigned right shift by 0 is idempotent - valid svalue
+
+pre:
+  - r0.type=number
+  #  This is 0x7fffffffffffffff, which is in range for a signed 64-bit number.
```
Comment (dthaler) [inline test-data/shift.yaml:1151]: the latter "which is in range..." phrase can be removed as it doesn't add any value in my opinion.

File: test-data/shift.yaml [hunk 4]
```diff
@@ -1104,3 +1104,167 @@ post:
...
+
+messages:
+  - "1:2: Code is unreachable after 1:2"
+
+---
+test-case: check that unsigned right shift by 0 is idempotent - svalue out of range - via register
+
+pre:
+  - r0.type=number
+  #  This is 0x8000000000000000, which is out of range for a signed 64-bit number.
```
Comment (dthaler) [inline test-data/shift.yaml:1189]: update comment here and in 1185

File: test-data/shift.yaml [hunk 5]
```diff
@@ -1104,3 +1104,167 @@ post:
...
+
+messages:
+  - "1:2: Code is unreachable after 1:2"
+
+---
+test-case: check that unsigned right shift by 0 is idempotent - valid svalue
+
+pre:
+  - r0.type=number
+  #  This is 0x7fffffffffffffff, which is in range for a signed 64-bit number.
```
Comment (dthaler) [inline test-data/shift.yaml:1233]: same comment as on line 1151

File: src/crab/ebpf_domain.cpp [hunk 6]
```diff
@@ -2464,11 +2459,9 @@ void ebpf_domain_t::ashr(const Reg& dst_reg, const linear_expression_t& right_sv
...
             }
             m_inv.set(dst.svalue, interval_t{lb_n, ub_n});
-            if (to_unsigned(ub_n) >= to_unsigned(lb_n)) {
-                m_inv.assign(dst.uvalue, dst.svalue);
-            } else {
-                havoc(dst.uvalue);
-            }
+            m_inv.assign(dst.uvalue, dst.svalue);
+            overflow_signed(m_inv, dst.svalue, finite_width);
+            overflow_unsigned(m_inv, dst.uvalue, finite_width);
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply src/crab/ebpf_domain.cpp:2464]: This path operates on the svalue, so assigning svalue -> uvalue is correct here.

**Review by dthaler — APPROVED** [review]

**Pull request merged**